netdev2_tap: assert address len >= ETHERNET_ADDR_LEN

cgundogan opened the pull request:
Using the shell: `ifconfig 4 set addr 11:22:33` crashes on native, because an assert is hit that expects the addr length to be `ETHERNET_ADDRESS_LEN`. As we already can provide return values for netopt set operations, I propose to change the `assert` to an actual `if` statement.
Currently, compiling without `asserts` would still allow to specify addresses with `length < ETHERNET_ADDRESS_LEN` and hence may lead to unexpected issues (random bytes in the address).

EDIT: assert address length to be `>= ETHERNET_ADDRESS_LEN`

kaspar030 commented:
I would rather have ifconfig check for a valid address length first, and do the check itself.
Otherwise, why check here for correct address length, but do asserts for most other vales?

cgundogan commented:
>I would rather have ifconfig check for a valid address length first, and do the check itself.
Otherwise, why check here for correct address length, but do asserts for most other vales?

The problem is that ifconfig does not know the limitations of the driver. I looked into the `at86rf` driver and there the lengh is handled similarly as proposed here. But on the other hand, other drivers (seems like most of them) use the `assert` method..

miri64 commented:
> I would rather have ifconfig check for a valid address length first, and do the check itself.

That would make the checks rather complicated, since this is very device dependent. Also for array values it is recommended to use `< MAX_LEN` in netdev anyway. So why not use that for the `assert()`? That would also still keep it in line with the other values and drivers.

cgundogan commented:
>Also for array values it is recommended to use < MAX_LEN in netdev anyway.

@miri64 can you elaborate on this? In the `cc2420` driver e.g. the asserts also check for the specific length of the address: `assert(val_len == 2);`.

I have no strong opinion about this anymore. We can close this PR if nobody has a better solution. But keep in mind, that native would still crash when a shorter address is supplied via the `ifconfig` shell command.

miri64 commented:
> @miri64 can you elaborate on this? In the cc2420 driver e.g. the asserts also check for the specific length of the address: assert(val_len == 2);.

If I see this correctly this isn't done very consistently through-out the code.

> I have no strong opinion about this anymore. We can close this PR if nobody has a better solution. But keep in mind, that native would still crash when a shorter address is supplied via the ifconfig shell command.

I just want you to replace the `!=` with a `<`. No need to close this PR.

kaspar030 commented:
> I just want you to replace the != with a <. No need to close this PR.

Actually we discussed this many times, and I AFAIR we concluded that testing for correct length at runtime is a waste. The caller needs to get the acceptable length and handle errors. In this case, the fix would be for the ifconfig shell command to get the address length before trying to set (a possibly wrong) one.

cgundogan commented:
> I just want you to replace the != with a <. No need to close this PR.

I don't understand what changing to `<` is supposed to fix here. If we go this route, then I would still prefer the current `==`, as any address with length less than `ETHERNET_ADDR_LEN` is an error.

> The caller needs to get the acceptable length and handle errors. 

@kaspar030 I agree with you, but as @miri64 stated before, the shell command needs a lot more logic to be able to discern the correct lengths from the devices in use.
Like I said before, I have no strong opinion regarding this. But it is still very *odd* and user unfriendly when an application crashes, because a user mistyped the address.

kaspar030 commented:
> @kaspar030 I agree with you, but as @miri64 stated before, the shell command needs a lot more logic to be able to discern the correct lengths from the devices in use.

Why a lot more logic? One call to get the needed length, one comparison. Probably a short table mapping address netopt number to address length netopt number.

miri64 commented:
> Why a lot more logic? One call to get the needed length, one comparison. Probably a short table mapping address netopt number to address length netopt number.

See there is where I see the waste / tricky handling. Why not just copy for array values (like addresses) when it fits and not just when it *exactly fits*. If you still want to crash for a user input why not `assert(addr >= ETHERNET_ADDRESS_LEN)`?

kaspar030 commented:
> See there is where I see the waste / tricky handling.

Isn't that just like 5 lines of code?

> Why not just copy for array values (like addresses) when it fits and not just when it exactly fits. If you still want to crash for a user input why not assert(addr >= ETHERNET_ADDRESS_LEN)?

Let's not crash on user input!

miri64 commented:
> Let's not crash on user input!

You were the one adding the asserts... But let me rephrase that: if you prefer assert over if, that I would prefer `assert(max_len >= ETHERNET_ADDRESS_LEN)`.

OlegHahm commented:
It's only [native after all](https://www.youtube.com/watch?v=PXYeARRyDWk).

cgundogan commented:
>But let me rephrase that: if you prefer assert over if, that I would prefer assert(max_len >= ETHERNET_ADDRESS_LEN)

@miri64 addressed your comment

kYc0o commented:
It's failing because `ETHERNET_ADDR_LEN` is not defined...

cgundogan commented:
>It's failing because ETHERNET_ADDRESS_LEN is not defined...

addressed
